src/agent/handlers/nodesHolePunchRequest.ts

```diff
@@ -0,0 +1,38 @@
+import type Logger from '@matrixai/logger';
+import type { HolePunchRequestMessage } from './types';
+import type { AgentRPCRequestParams, AgentRPCResponseResult } from '../types';
+import type NodeConnectionManager from '../../nodes/NodeConnectionManager';
+import type { Host, Port } from '../../network/types';
+import * as agentErrors from '../errors';
+import * as agentUtils from '../utils';
+import { UnaryHandler } from '../../rpc/handlers';
+
+class NodesHolePunchRequestHandler extends UnaryHandler<
+  {
+    nodeConnectionManager: NodeConnectionManager;
+    logger: Logger;
+  },
+  AgentRPCRequestParams<HolePunchRequestMessage>,
+  AgentRPCResponseResult
+> {
+  public async handle(
+    input: AgentRPCRequestParams<HolePunchRequestMessage>,
+    _cancel,
+    meta,
+  ): Promise<AgentRPCResponseResult> {
+    const { nodeConnectionManager, logger } = this.container;
+    // Connections should always be validated
+    const requestingNodeId = agentUtils.nodeIdFromMeta(meta);
+    if (requestingNodeId == null) {
+      throw new agentErrors.ErrorAgentNodeIdMissing();
+    }
```

Comment:
I don't see how this accomplishes #148. The receiver is supposed to verify that the message is signed by the original sending node?

Reply:
I opted not to do that in the end, I concluded it didn't add any real functionality or security. I can add it it, but it won't prevent other nodes from requesting punches to a target node. Mostly because we can verify the the information has been signed by someone but we can't verify that the information is correct.

A bad actor can just generate a signed request to some other address.

Reply:
As discussed, it's still better to have this extra layer of protection, using the `Token` class should be sufficient.

And I want to leave the possibility of extensibility here with further crypto protection in the future.

Reply:
I've added in message signing and verification.